docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,363 @@
+
+`<argocd-host>/api/v1/extensions/<extension-name>`
+
+With the configuration bellow, the expected behavior is explained in the
+following examples:
+
+```yaml
+extension.config: |
+  extensions:
+    - name: some-extension
+      enabled: true
+      backend:
+        idleConnTimeout: 10s
+        services:
+          - url: http://extension-name.com:8080
+```
+
+- **Example 1**:
+
+Argo CD API server acts as a reverse-proxy forwarding http requests as
+follows:
+
+```
+   ┌────────────┐
+   │ Argo CD UI │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension
+          │
+          ▼
+ ┌──────────────────┐
+ │Argo CD API Server│
+ └────────┬─────────┘
+          │
+          │ GET http://extension-name.com:8080
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+- **Example 2**:
+
+If a backend provides an API under the `/apiv1/metrics` endpoint, Argo CD
+should be able to invoke it such as:
+
+```
+   ┌────────────┐
+   │ Argo CD UI │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension/apiv1/metrics/123
+          │
+          ▼
+ ┌──────────────────┐
+ │Argo CD API Server│
+ └────────┬─────────┘
+          │
+          │ GET http://extension-name.com:8080/apiv1/metrics/123
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+- **Example 3**:
+
+In this use-case we have one Argo CD instance connected with different
+clusters. There is a requirement defining that every extension instance
+needs to be deployed in each of the target clusters. To address this
+use-case there is a need to configure multiple backend URLs for the
+same extension (one for each cluster). For doing so, the following
+configuration should be possible:
+
+```yaml
+extension.config: |
+  extensions:
+    - name: some-extension
+      enabled: true
+      backend:
+        idleConnTimeout: 10s
+        services:
+          - url: http://extension-name.com:8080
+            clusterName: kubernetes.local
+          - url: https://extension-name.ppd.cluster.k8s.local:8080
+            clusterName: admins@ppd.cluster.k8s.local
+```
+
+Note that there is an URL configuration per cluster name. The cluster
+name is extracted from the Argo CD cluster secret and must match the
+field `data.name`. In this case the UI must send a header with the
+application name so API Server can check in which cluster it should
+get the backend URL from. This will be done by inspecting the
+Application destination configuration to find the proper cluster
+name.
+
+The diagram below shows how Argo CD UI could send the request with
+the additional header to get the proxy forwarding it to the proper
+cluster:
+
+```
+   ┌────────────┐
+   │ Argo CD UI │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension
+          │ HEADER: "ARGOCD-APPLICATION-NAME: ppd-application
```

Comment:
This doc should probably specify that the header must include the _fully-qualified_ namespace/appname name.

If a user is permitted to perform the action on `default/my-app` but not on `admin/my-app`, then we must make sure the backend is able to differentiate the two.

Reply:
Makes sense. Will update the proposal.

Reply:
updated